src/snowflake/snowpark/mock/_nop_connection.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from logging import getLogger
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import pandas
+
+from snowflake.connector.cursor import ResultMetadata
+from snowflake.snowpark._internal.analyzer.analyzer_utils import unquote_if_quoted
+from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import (
+    LogicalPlan,
+    SaveMode,
+    SnowflakeCreateTable,
+)
+from snowflake.snowpark._internal.analyzer.table_merge_expression import TableUpdate
+from snowflake.snowpark._internal.analyzer.unary_plan_node import CreateViewCommand
+from snowflake.snowpark._internal.server_connection import DEFAULT_STRING_SIZE
+from snowflake.snowpark._internal.utils import result_set_to_rows
+from snowflake.snowpark.async_job import _AsyncResultType
+from snowflake.snowpark.mock._connection import MockServerConnection
+from snowflake.snowpark.mock._nop_plan import NopExecutionPlan
+from snowflake.snowpark.mock._telemetry import LocalTestOOBTelemetryService
+from snowflake.snowpark.mock._util import get_fully_qualified_name
+from snowflake.snowpark.row import Row, canonicalize_field
+from snowflake.snowpark.types import (
+    ArrayType,
+    BinaryType,
+    BooleanType,
+    DateType,
+    GeographyType,
+    GeometryType,
+    MapType,
+    NullType,
+    StringType,
+    StructType,
+    TimestampType,
+    TimeType,
+    VariantType,
+    _NumericType,
+)
+
+if TYPE_CHECKING:
+    try:
+        from snowflake.connector.cursor import ResultMetadataV2
+    except ImportError:
+        ResultMetadataV2 = ResultMetadata
+
+logger = getLogger(__name__)
+
+
+class NopConnection(MockServerConnection):
+    class NopEntityRegistry(MockServerConnection.TabularEntityRegistry):
+        def write_table(
+            self,
+            name: Union[str, Iterable[str]],
+            table: LogicalPlan,
+            mode: SaveMode,
+            column_names: Optional[List[str]] = None,
+        ) -> List[Row]:
+            current_schema = self.conn._get_current_parameter("schema")
+            current_database = self.conn._get_current_parameter("database")
+            name = get_fully_qualified_name(name, current_schema, current_database)
+            self.table_registry[name] = table
+            return [Row(status=f"Table {name} successfully created.")]
+
+    def __init__(self, options: Optional[Dict[str, Any]] = None) -> None:
+        super().__init__(options)
+        self._conn._session_parameters = {
```

Comment:
nit: _session_parameters should be Dict[str,str], at least many APIs use that as type hint.

Reply:
MockServerConnection defines as Dict[str, Any] and I checked a few places the retrieval code assumes the value is a boolean.

Reply:
that sounds like something out of scope for this PR, but we prob. should consider fixing in the existing API. So basically the `statement_params` is not uniquely typed. Ideally in the AST we would represent this with one object/type.